### PR TITLE
Query Analysis first attempt: Cardinality & Generic Cost

### PR DESCRIPTION
**Pull request by rabbitonweb**
Part of #1716

### CODE REVIEW

**Comment by rabbitonweb** [comment]
You should see this comment in Slack....

File: connector/src/main/scala/quasar/qanalysis/package.scala [hunk 1]
```diff
@@ -0,0 +1,149 @@
+/*
```
Comment (alissapajer) [inline connector/src/main/scala/quasar/qanalysis/package.scala]: Testing PR review comment for Slack.

File: connector/src/main/scala/quasar/qanalysis/package.scala [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+ * Copyright 2014–2017 SlamData Inc.
```
Comment (alissapajer) [inline connector/src/main/scala/quasar/qanalysis/package.scala]: Testing PR review comment for Slack take 2.

File: connector/src/main/scala/quasar/qanalysis/package.scala [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+ *
```
Comment (alissapajer) [inline connector/src/main/scala/quasar/qanalysis/package.scala]: test comment 3

**Comment by rabbitonweb** [comment]
Missing elements (will be part of separate PR):

1. change return type of `Cardinality` to return range `(Int, Option[Int]) where `fst` is minimal and `snd` is `None` if unknown or Some(max)
2. change `pathCard` to return `M[Int]` instead of `Int` where `M : Monad`

**Comment by rabbitonweb** [comment]
@sellout pls review

File: connector/src/main/scala/quasar/qanalysis/package.scala [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    implicit def qscriptCore[T[_[_]]: RecursiveT: ShowT]: Cost[QScriptCore[T, ?]] =
+      new Cost[QScriptCore[T, ?]] {
+        def evaluate(pathCard: APath => Int): GAlgebra[(Int, ?), QScriptCore[T, ?], Int] = {
+          case qscript.Map((card, cost), f) => card + cost
+          case Reduce((card, cost), bucket, reducers, repair) => card + cost
+          case Sort((card, cost), bucket, orders) => card + cost
+          case Filter((card, cost), f) => card + cost
+          case Subset((card, cost), from, sel, count) => card + cost
+          case LeftShift((card, cost), struct, id, repair) => card + cost
+          case Union((card, cost), lBranch, rBranch) => card + cost
```
Comment (sellout) [inline connector/src/main/scala/quasar/qanalysis/package.scala]: This case should look like the Join cases.

File: connector/src/main/scala/quasar/qanalysis/package.scala [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+  }
+
+  @typeclass
+  trait Analyzer[F[_]] {
+    def analyze[T](t: T)(pathCard: APath => Int)(implicit
+      rec: Recursive.Aux[T, F],
+      func: Functor[F],
+      cardinalty: Cardinality[F],
+      cost: Cost[F]
+    ): Int = rec.zygo(t)(cardinalty.calculate(pathCard), cost.evaluate(pathCard))
```
Comment (sellout) [inline connector/src/main/scala/quasar/qanalysis/package.scala]: Actually, why is this even in a type class? You can just expose the function directly.

File: connector/src/main/scala/quasar/qanalysis/package.scala [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+  case object Unknown extends AnalysisResult
+
+  @typeclass
+  trait Cardinality[F[_]] {
+    def calculate(pathCard: APath => Int): Algebra[F, Int]
+  }
+
+  @typeclass
+  trait Cost[F[_]] {
+    def evaluate(pathCard: APath => Int): GAlgebra[(Int, ?), F, Int]
```
Comment (rabbitonweb) [inline connector/src/main/scala/quasar/qanalysis/package.scala]: @sellout `evaluate` needs `pathCard` becaue when we deal with `EquiJoin` and `ThetaJoin` we need to call `zygo` on both left and right branch

**Comment by rabbitonweb** [comment]
@sellout i will go with range in separate PR. I don't think I will put any probabilistic for `4.2`

**Comment by rabbitonweb** [comment]
@sellout this is ready for review

**Comment by sellout** [comment]
@rabbitonweb There are still a bunch of comments you haven’t addressed.

File: connector/src/main/scala/quasar/qanalysis/package.scala (hunk 5 above)
Comment (rabbitonweb) [inline connector/src/main/scala/quasar/qanalysis/package.scala]: Why is `analyze` a type class? I was thinking that we could have a single entry point for connector creators to use

Comment (sellout) [inline connector/src/main/scala/quasar/qanalysis/package.scala]: A single entry point how? Do you expect connectors to override `analyze`? It definitely doesn’t need to be parameterized on `F`, as if it is, you’ll need to create empty(?) instances for each component of the coproduct.

**Comment by sellout** [comment]
@rabbitonweb Just want to make sure you’re aware this build isn’t passing.

**Pull request merged**